File: src/Functional/Haskell/patterns/Strategy.hs
module Main where

type PricingStrategy = Int -> Int

price :: Int -> PricingStrategy -> Int
price value strategy = strategy value

regular :: PricingStrategy
regular value = value

vip :: PricingStrategy
vip value = value * 80 `div` 100

main :: IO ()
main =
  if price 100 regular == 100 && price 100 vip == 80
    then putStrLn "regular=100;vip=80"
    else error "Strategy contract failed"

File: src/Systems/Zig/patterns/strategy.zig
const std = @import("std");

const PricingStrategy = *const fn (i32) i32;

fn price(value: i32, strategy: PricingStrategy) i32 {
    return strategy(value);
}

fn regular(value: i32) i32 {
    return value;
}

fn vip(value: i32) i32 {
    return @divTrunc(value * 80, 100);
}

pub fn verifyStrategy() bool {
    return price(100, regular) == 100 and price(100, vip) == 80;
}

pub fn main() !void {
    if (!verifyStrategy()) {
        return error.StrategyContractFailed;
    }
    std.debug.print("regular=100;vip=80\n", .{});
}
